### PR TITLE
add support for nastympi in dart-mpi

### PR DESCRIPTION
**Pull request by fmoessbauer**
Refers to #52.

### CODE REVIEW

**Comment by fuchsto** [comment]
Awesome! But this does not use NastyMPI in CI yet, right?
NastMPI is maintained at github in an official DASH repository:

https://github.com/dash-project/nasty-MPI

So if the user enables it via flag, CMake should pull and build it, just like we do with GoogleTest.
It's a tiny lib and I'm quite sure it doesn't add any new dependencies.

Also, the log file of CircleCI/mpich ends in:

```
[[ RUN    ]] mpirun  -n 1  /opt/dash/build-ci/20161105-203754/Minimal/bin/dash/test/mpi/dash-test-mpi
[proxy:0:0@e0da45fb18a9] HYDU_create_process (utils/launch/launch.c:75): execvp error on file /opt/dash/build-ci/20161105-203754/Minimal/bin/dash/test/mpi/dash-test-mpi (No such file or directory)
[[     OK ]] Tests passed, returned 0
[[== FINISHED =================================================]]
[[== START ====================================================]]
[[ RUN    ]] mpirun  -n 2  /opt/dash/build-ci/20161105-203754/Minimal/bin/dash/test/mpi/dash-test-mpi
[proxy:0:0@e0da45fb18a9] HYDU_create_process (utils/launch/launch.c:75): execvp error on file /opt/dash/build-ci/20161105-203754/Minimal/bin/dash/test/mpi/dash-test-mpi (No such file or directory)
[proxy:0:0@e0da45fb18a9] HYDU_create_process (utils/launch/launch.c:75): execvp error on file /opt/dash/build-ci/20161105-203754/Minimal/bin/dash/test/mpi/dash-test-mpi (No such file or directory)
[[     OK ]] Tests passed, returned 0
[[== FINISHED =================================================]]
[[== START ====================================================]]
[[ RUN    ]] mpirun  -n 3  /opt/dash/build-ci/20161105-203754/Minimal/bin/dash/test/mpi/dash-test-mpi
[proxy:0:0@e0da45fb18a9] HYDU_create_process (utils/launch/launch.c:75): execvp error on file /opt/dash/build-ci/20161105-203754/Minimal/bin/dash/test/mpi/dash-test-mpi (No such file or directory)
[proxy:0:0@e0da45fb18a9] HYDU_create_process (utils/launch/launch.c:75): execvp error on file /opt/dash/build-ci/20161105-203754/Minimal/bin/dash/test/mpi/dash-test-mpi (No such file or directory)
[proxy:0:0@e0da45fb18a9] HYDU_create_process (utils/launch/launch.c:75): execvp error on file /opt/dash/build-ci/20161105-203754/Minimal/bin/dash/test/mpi/dash-test-mpi (No such file or directory)
[[     OK ]] Tests passed, returned 0
[[== FINISHED =================================================]]
[     OK ]
[ PASSED ] Build and test suite passed
cp: cannot stat ‘./build-ci/*/Minimal/dash-tests-*.xml’: No such file or directory
checking **logs**
```

It tries to run the minimal build apparently.

**Comment by fmoessbauer** [comment]
You're right. This PR just enables DASH's build system to use NastyMPI. What you said is exactly what I said in #52, but currently this is not easily possible as NastyMPI uses make with hardcoded pathes. For a seamless integration cmake support of NastyMPI would be nice. Assigned it to Roger.

The errors in the log are there since PR #59. A tiny bug from some merge disabled the tests in the minimal target, hence mpirun does not find the binary. Unfortunately I did not detect that in the PR.
This is fixed in #65, which cannot be merge as there are still some bugs in DASH if building without locality support packages (hwloc, numa, ...).

## In Summary:
- You can safely merge the PR and I will make another for NastyMPI in CI.
- We need some project management to keep track of related issues.

**Comment by fmoessbauer** [comment]
All points @fuchsto mentioned are implemented now. Currently NastyMPI is build using the provided makefile. This is ok, until we have cmake support in NastyMPI.

@rkowalewski Please verify that NastyMPI is correctly integrated in DASH / DART-MPI. Is there anything more to do but linking agains?

**Comment by fuchsto** [comment]
@fmoessbauer
Actually NastyMPI is not linked but used as dynamic library at runtime via `LD_PRELOAD`.

Static linking should work, too. 
On some systems like Cray, dynamic linking is not available.
The link order is highly relevant in this case. It must be ensured that MPI functions are answered by NastyMPI and not by the original MPI library. 

Is this ensured in some way?

**Comment by fuchsto** [comment]
Squashed and merged, but leaving branch `feat-52-nastympi` open for further changes.
